quad_utils/config/filter_chain.yaml

```diff
@@ -85,36 +61,65 @@ grid_map_filters:
       output_layer: roughness
       expression: abs(z_inpainted - z_smooth)
 
-  # Edge detection by computing the standard deviation from slope.
-  - name: edge_detection
-    type: gridMapFilters/SlidingWindowMathExpressionFilter
-    params:
-      input_layer: slope
-      output_layer: edges
-      expression: sqrt(sumOfFinites(square(slope - meanOfFinites(slope))) ./ numberOfFinites(slope)) # Standard deviation
-      compute_empty_cells: false
-      edge_handling: crop # options: inside, crop, empty, mean
-      window_length: 0.2
-
-  # # Edge detection on z layer with convolution filter as alternative to filter above.
+  # # Uncomment to use edges
+  # # Edge detection by computing the standard deviation from slope.
   # - name: edge_detection
   #   type: gridMapFilters/SlidingWindowMathExpressionFilter
   #   params:
-  #     input_layer: z_inpainted
+  #     input_layer: slope
   #     output_layer: edges
-  #     expression: 'sumOfFinites([0,1,0;1,-4,1;0,1,0].*z_inpainted)' # Edge detection.
-  #     # expression: 'sumOfFinites([0,-1,0;-1,5,-1;0,-1,0].*z_inpainted)' # Sharpen.
+  #     expression: sqrt(sumOfFinites(square(slope - meanOfFinites(slope))) ./ numberOfFinites(slope)) # Standard deviation
   #     compute_empty_cells: false
-  #     edge_handling: mean # options: inside, crop, empty, mean
-  #     window_size: 3 # Make sure to make this compatible with the kernel matrix.
+  #     edge_handling: crop # options: inside, crop, empty, mean
+  #     window_length: 0.2
 
   # Compute traversability as normalized weighted sum of slope and roughness.
   - name: traversability
     type: gridMapFilters/MathExpressionFilter
     params:
-      output_layer: traversability
+      output_layer: traversability_raw
       expression: 0.5 * (1.0 - (slope / 0.3)) + 0.5 * (1.0 - (roughness / 0.4))
 
+  #   # Compute traversability as normalized weighted sum of slope and roughness.
+  # - name: traversability_user_mask
+  #   type: gridMapFilters/MathExpressionFilter
+  #   params:
+  #     output_layer: traversability
+  #     expression: ( (0.5 - 0.5 * (x - 3.0) ./ (sqrt((x - 3.0).^2))) + 
+  #                   (0.5 + 0.5 * (x - 3.2) ./ (sqrt((x - 3.2).^2))) + 
+  #                   (0.5 - 0.5 * (y + 1.0) ./ (sqrt((y + 1.0).^2))) + 
+  #                   (0.5 + 0.5 * (y - 1.0) ./ (sqrt((y - 1.0).^2))))
+  #                 .* traversability
+
+  # Set lower threshold on traversability.
+  - name: z_finite_threshold
+    type: gridMapFilters/ThresholdFilter
+    params:
+      condition_layer: z
+      output_layer: z_finite
+      lower_threshold: -1000.0
+      set_to: -1000.0
+
+  # Compute traversability as normalized weighted sum of slope and roughness.
+  - name: traversability_user_mask
+    type: gridMapFilters/MathExpressionFilter
+    params:
+      output_layer: traversability_raw
+      expression: (1 - abs(z_finite - z_inpainted)) .* traversability_raw
```

Comment:
Where are you using the mask? And what is the value assigned to the gap area for z_finite and z_inpainted ? If there are no "nans" in the grid I don't think the inpaint filter will change anything.

Reply:
Yeah this is certainly up for refactoring to something that makes more sense but it does work. The map does actually have holes (nans) so in z_finite those are set to a low number (-1000). This makes (1 - abs(z_finite - z_inpainted)) = 1 if the inpainting did nothing (i.e. data already existed there) and a very negative number if it did not (i.e. untraversable).